ada_feeding_perception/ada_feeding_perception/face_detection.py

```diff
@@ -0,0 +1,388 @@
+                    ),
+                ),
+            ],
+        )
+
+    def toggle_face_detection_callback(
+        self, request: SetBool.Request, response: SetBool.Response
+    ) -> SetBool.Response:
+        """
+        Callback function for the toggle_face_detection service. Safely toggles
+        the face detection on or off depending on the request.
+        """
+        self.get_logger().info(f"Incoming service request. data: {request.data}")
+        with self.is_on_lock:
+            self.is_on = request.data
+        response.success = request.data
+        return response
+
+    def camera_info_callback(self, msg: CameraInfo):
+        """
+        Callback function for the toggle_face_detection service. Safely toggles
+        the face detection on or off depending on the request.
+        """
+        self.camera_matrix = msg.k
+
+    def depth_callback(self, msg: CompressedImage):
+        """
+        Callback function for depth images. If face_detection is on, this
+        function publishes the 3d location of the mouth on the /face_detection
+        topic in the camera_depth_optical_frame
+        """
+
+        # Collect last ten depth images (depth_buffer is a collections.deque of max length 10)
+        self.depth_buffer.append(msg)
+        with self.is_on_lock:
+            is_on = self.is_on
+
+        # Check if face detection is on
+        if is_on:
+            face_detection_msg = FaceDetection()
+            if self.is_face_detected:
+                # Retrieve the 2d location of the mouth center
+                u = int(self.img_mouth_center[0])
+                v = int(self.img_mouth_center[1])
+
+                # Find depth image closest in time to saved color image
+                difference_array = np.array([])
+                for depth_img in self.depth_buffer:
+                    depth_time = depth_img.header.stamp.sec + (
+                        depth_img.header.stamp.nanosec / (10**9)
+                    )
+                    img_time = self.img_timestamp.sec + (
+                        self.img_timestamp.nanosec / (10**9)
+                    )
+                    difference_array = np.append(
+                        difference_array, abs(depth_time - img_time)
+                    )
+                closest_depth = self.bridge.imgmsg_to_cv2(
+                    self.depth_buffer[difference_array.argmin()],
+                    desired_encoding="passthrough",
+                )
+
+                # Retrieve the depth value averaged over all mouth coordinates
+                depth_sum = 0
+                for point in self.img_mouth_points:
+                    depth_sum += closest_depth[int(point[1])][int(point[0])]
+                depth = depth_sum / float(len(self.img_mouth_points))
+
+                # Create target 3d point, with mm measurements converted to m
+                mouth_location = PointStamped()
+                mouth_location.header = msg.header
```

Comment:
I'd actually prefer if the header is the image message header as opposed to the depth message header, so that there is a one-to-one correspondence between the headers in the annotated image message and the FaceDetection message.  Since you already have the img_timestamp, it is just a matter of `mouth_location.header = Header(stamp=img_timestamp)`